### PR TITLE
[llvm] Use -place-safepoints in JIT mode too

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes: https://github.com/mono/mono/issues/17059

Example:
```csharp
static int Test()
{
    int sum = 0;
    for (int i = 0; i < 42000; i++)
        sum++;
    return sum;
}
```
### Current behavior:
```llvm
define monocc i32 @"Program:Test ()"() #0 {
BB0:
  %const = bitcast i64 140647394572600 to i64
  %0 = inttoptr i64 %const to i64*
  %1 = load volatile i64, i64* %0, align 8
  %2 = icmp eq i64 %1, 0
  br i1 %2, label %BB2, label %POLL_BB1, !prof !1

POLL_BB1:                                         ; preds = %BB0
  %3 = load void ()*, void ()** @mono_threads_state_poll, align 8
  call coldcc void %3()
  br label %BB2

BB2:                                              ; preds = %POLL_BB1, %BB0
  br label %BB4

BB4:                                              ; preds = %CONT_BB4, %BB2
  %lsr.iv = phi i32 [ %lsr.iv.next, %CONT_BB4 ], [ 42000, %BB2 ]
  %exitcond = icmp eq i32 %lsr.iv, 0
  br i1 %exitcond, label %BB1, label %BB5

BB5:                                              ; preds = %BB4
  %4 = inttoptr i64 %const to i64*
  %5 = load volatile i64, i64* %4, align 8
  %6 = icmp eq i64 %5, 0
  br i1 %6, label %CONT_BB4, label %POLL_BB3, !prof !1

POLL_BB3:                                         ; preds = %BB5
  %7 = load void ()*, void ()** @mono_threads_state_poll.1, align 8
  call coldcc void %7()
  br label %CONT_BB4

CONT_BB4:                                         ; preds = %POLL_BB3, %BB5
  %lsr.iv.next = add nsw i32 %lsr.iv, -1
  br label %BB4

BB1:                                              ; preds = %BB4
  ret i32 42000
}
```
(loop was not optimized away due to early inserted safepoints)

### New behavior:
```llvm
define monocc i32 @"Program:Test ()"() #0 gc "coreclr" {
BB0:
  %0 = load volatile i64, i64* inttoptr (i64 139875651025208 to i64*)
  %1 = icmp eq i64 %0, 0
  br i1 %1, label %gc.safepoint_poll.exit, label %gc.safepoint_poll.poll.i, !prof !1

gc.safepoint_poll.poll.i:
  %2 = load void ()*, void ()** @mono_threads_state_poll
  call coldcc void %2() #1
  br label %gc.safepoint_poll.exit

gc.safepoint_poll.exit: 
  ret i32 42000 ;;; NOTE: return 42000!
}
```
(LLVM itself put safepoints after optimizations so the loop was optimized!)

Should significantly improve LLVM JIT performance in benchmarks (in non-preemptive mode)

### CODE REVIEW

**Comment by EgorBo** [comment]
Oops, it supposed to be a draft PR.

File: mono/mini/llvm-jit.cpp [hunk 1]
```diff
@@ -284,13 +285,14 @@ class MonoLLVMJIT {
...
 		initializeTarget(registry);
+		linkCoreCLRGC();
 
 		const char *opts = g_getenv ("MONO_LLVM_OPT");
 		if (opts == NULL) {
 			// FIXME: find optimal mono specific order of passes
 			// see https://llvm.org/docs/Frontend/PerformanceTips.html#pass-ordering
 			// the following order is based on a stripped version of "OPT -O2"
-			opts = " -simplifycfg -sroa -lower-expect -instcombine -licm -simplifycfg -lcssa -indvars -loop-deletion -gvn -memcpyopt -sccp -bdce -instcombine -dse -simplifycfg";
+			opts = " -simplifycfg -sroa -lower-expect -instcombine -licm -simplifycfg -lcssa -indvars -loop-deletion -gvn -memcpyopt -sccp -bdce -instcombine -dse -simplifycfg -place-safepoints";
```
Comment (filipnavara) [inline mono/mini/llvm-jit.cpp]: Don't we need to add `-place-safepoints` unconditionally irrespective of `MONO_LLVM_OPT`?

Comment (EgorBo) [reply mono/mini/llvm-jit.cpp]: Makes sense! Fixed

File: mono/mini/llvm-jit.cpp [hunk 2]
```diff
@@ -284,26 +285,34 @@ class MonoLLVMJIT {
...
-			// see https://llvm.org/docs/Frontend/PerformanceTips.html#pass-ordering
-			// the following order is based on a stripped version of "OPT -O2"
-			opts = " -simplifycfg -sroa -lower-expect -instcombine -licm -simplifycfg -lcssa -indvars -loop-deletion -gvn -memcpyopt -sccp -bdce -instcombine -dse -simplifycfg";
-		}
+		if (opts == NULL)
+			opts = default_opts;
+		else if (opts[0] == '+') // Append passes to the default order if starts with '+', overwrite otherwise
+			opts = g_strdup_printf ("%s %s", default_opts, opts + 1);
+		else if (opts[0] != ' ') // pass order has to start with a leading whitespace
+			opts = g_strdup_printf (" %s", opts);
```
Comment (EgorBo) [inline mono/mini/llvm-jit.cpp]: Unrelated change, just improved `MONO_LLVM_OPT` a bit (it's now possible to append passes to the default order instead of completely overwrite them.

File: mono/mini/mini-llvm.c [hunk 3]
```diff
@@ -4676,6 +4676,56 @@ get_float_const (MonoCompile *cfg, float val)
 		return LLVMConstFPExt (LLVMConstReal (LLVMFloatType (), val), LLVMDoubleType ());
 }
 
+static void
+emit_safepoint_poll_jit (EmitContext *ctx)
+{
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: Can't this be merged with the aot version ?

File: mono/mini/mini-llvm.c [hunk 4]
```diff
@@ -8118,8 +8168,11 @@ emit_method_inner (EmitContext *ctx)
 		}
 	}
 
-	if (!cfg->llvm_only && cfg->compile_aot && mono_threads_are_safepoints_enabled () && requires_safepoint)
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: The OP_GC_SAFE_POINT code should be removed.

Comment (filipnavara) [reply mono/mini/mini-llvm.c]: Still used for the LLVM-only mode.

Comment (EgorBo) [reply mono/mini/mini-llvm.c]: Yeah I initially removed it but then Filip reminded me the llvm-only case

File: mono/mini/mini-llvm.c (hunk 3 above)
Comment (EgorBo) [reply mono/mini/mini-llvm.c]: Ok, will try

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
Also, wouldn't this increase memory usage because every method will have a safepoint function emitted ?

**Comment by filipnavara** [comment]
@vargaz I thought so but apparently it gets inlined and the result is not worse than it was before. I didn't run it locally to prove this but the theory seemed plausible.

**Comment by vargaz** [comment]
I mean every jit module is going to have that extra unused function, or does it get removed ?

**Comment by EgorBo** [comment]
@vargaz I've just pushed a change,
now the function is deleted after optimizations (`-place-safepoints` is the last pass) and should not affect codegen size/time/memory anyhow.

Also, I've put "optnone" attribute on it so the optimizer doesn't need to waste time on it.

**Comment by EgorBo** [comment]
Here is a diff for the change: https://www.diffchecker.com/DEiZVPkP
(Full Module IR after optimizations), the function is deleted.

**Comment by EgorBo** [comment]
@filipnavara honestly, no idea why loaded llvm doesn't work (it failed with "symbol not found" something like that) I thought "loaded llvm" mode was removed.

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build Linux ARMv7

**Comment by EgorBo** [comment]
hm.. System.Runtime.Tests fail at some point with 
```
Cannot transition thread 0x7fa3af7b0700 from SELF_SUSPENDED with REQUEST_PULSE
```

**Comment by lambdageek** [comment]
@EgorBo if this happens again, could you grab the stack trace - that's very suspicious it seems like a race condition somewhere
> `Cannot transition thread 0x7fa3af7b0700 from SELF_SUSPENDED with REQUEST_PULSE`

**Comment by EgorBo** [comment]
@lambdageek it happens always in my branch 🙁 
stack-traces:
https://gist.github.com/EgorBo/d1b4caf4c92bf17dd8b6cb9d5a1fefad
or
https://gist.github.com/EgorBo/d6be13a0f8d12a2af0d90a090dc7355e

Maybe I emit safepoints where I should not?

**Comment by EgorBo** [comment]
Smaller repro:
```csharp
using System;
using System.Threading;

namespace HelloWorld
{
    class Program
    {
        static void Main(string[] args)
        {
            ThreadPool.QueueUserWorkItem(_ => {
                GC.Collect();
            });

            for (int i = 0; i < 10; i++) {
                ThreadPool.QueueUserWorkItem(_ => {
                    while (true) {
                        var o = new object();
                    }
                });
            }

            Thread.Sleep(100000);
        }
    }
}
```

**Comment by EgorBo** [comment]
I suspect I should ignore `cfg->method->wrapper_type == MONO_WRAPPER_ALLOC` methods (it somehow helps with the repro but the tests still crash)

**Comment by EgorBo** [comment]
Fixed with `cfg->method->wrapper_type != MONO_WRAPPER_ALLOC` check (do not emit sp in allocators)

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by lewing** [comment]
Where are we with this?

**Comment by EgorBo** [comment]
@lewing the PR is finished.

**Comment by SamMonoRT** [comment]
@EgorBo  - all clear to merge ?
